File: src/797.py
class Solution:
    def allPathsSourceTarget(self, graph: list[list[int]]) -> list[list[int]]:
        result = []
        path = [0]
        target = len(graph)-1

        def dfs(node):
            if node == target:
                result.append(path[:])
                return

            for neib in graph[node]:
                path.append(neib)
                dfs(neib)
                path.pop()

        dfs(0)
        return result

File: src/797.kt
class Solution {
    private val result = mutableListOf<List<Int>>()

    fun allPathsSourceTarget(graph: Array<IntArray>): List<List<Int>> {
        val n = graph.size
        val path = mutableListOf(0)
        dfs(graph, 0, n - 1, path)
        return result
    }


    private fun dfs(graph: Array<IntArray>, from: Int, to: Int, path: MutableList<Int>){
        if(from == to){
            result.add(path.toList())
            return
        }
        for(vert in graph[from]){
            path.add(vert)
            dfs(graph, vert, to, path)
            path.removeAt(path.size - 1)
        }
    }
}
